src/services/user-group/hooks/courses.js

```diff
@@ -0,0 +1,120 @@
+			).exec();
+			return hook;
+		});
+	}
+	return hook;
+};
+
+/**
+ * deletes all students from a course when a class is removed from the course
+ * this function goes into a before hook before we have to check whether there is a class missing
+ * in the patch-body which was in the course before
+ * @param hook - contains and request body
+ */
+const deleteWholeClassFromCourse = (hook) => {
+	const requestBody = hook.data;
+	const courseId = hook.id;
+	if (requestBody.classIds === undefined && requestBody.user === undefined) {
+		return hook;
+	}
+	return CourseModel.findById(courseId).exec().then((course) => {
+		if (!course) return hook;
+
+		const removedClasses = _.differenceBy(course.classIds, requestBody.classIds, v => JSON.stringify(v));
+		if (removedClasses.length < 1) return hook;
+		return Promise.all(removedClasses.map(classId => ClassModel.findById(classId).exec().then(c => (c || []).userIds))).then(async (studentIds) => {
+			// flatten deep arrays and remove duplicates
+			studentIds = _.uniqWith(_.flattenDeep(studentIds), (e1, e2) => JSON.stringify(e1) === JSON.stringify(e2));
+
+			// remove class students from course DB and from hook body to not patch them back
+			await CourseModel.update(
+				{ _id: course._id },
+				{ $pull: { userIds: { $in: studentIds } } },
+				{ multi: true },
+			).exec();
+			hook.data.userIds = hook.data.userIds.filter(value => !studentIds.some(id => id.toString() === value));
+			return hook;
+		});
+	});
+};
+
+const courseInviteHook = async (context) => {
+	if (context.path === 'courses' && context.params.query && context.params.query.link) {
+		// link is used as "authorization"
+		const dbLink = await context.app.service('link').get(context.params.query.link);
+		delete context.params.query.link;
+		if (dbLink) return restrictToCurrentSchool(context);
+	}
+
+	return restrictToUsersOwnCourses(context);
+};
+
+const patchPermissionHook = async (context) => {
+	const query = context.params.query || {};
+	const defaultPermissionHook = ctx => Promise.resolve(globalHooks.hasPermission('USERGROUP_EDIT')(ctx))
+		.then(_ctx => restrictToUsersOwnCourses(_ctx));
+
+	if (query.link) {
+		const dbLink = await context.app.service('link').get(query.link); // link is used as "authorization"
+		delete context.params.query.link;
+		if (dbLink) return context;
+	}
+
+	return defaultPermissionHook(context);
+};
+
+/**
+ * If the course is expired (archived), only the untilDate may be changed.
+ * @param context contains the feathers context of the request
+ */
+const restrictChangesToArchivedCourse = async (context) => {
+	const course = await context.app.service('courses').get(context.id);
+	const oneDayInMilliseconds = 864e5;
+	if (course.untilDate === undefined || course.untilDate >= Date.now() - oneDayInMilliseconds) {
+		return context;
+	}
+	// course is expired
+	let allowed = true;
+	Object.keys(context.data).forEach(((key) => {
+		if (key !== 'untilDate' && key !== 'startDate') allowed = false;
```

Comment:
what happens here? the logic seems inverted and why not directly access the data properties?

Reply:
discussed in person and fixed